stemcell_builder/stages/image_aws_install_grub/apply.sh

```diff
@@ -0,0 +1,113 @@
+#umount ${image_mount_point} || true
+#losetup -j ${disk_image} | cut -d ':' -f 1 | xargs --no-run-if-empty losetup -d
+kpartx -dv ${disk_image}
+
+# note: if the above kpartx command fails, it's probably because the loopback device needs to be unmapped.
+# in that case, try this: sudo dmsetup remove loop0p1
+
+# Map partition in image to loopback
+device=$(losetup --show --find ${disk_image})
+add_on_exit "losetup --verbose --detach ${device}"
+
+device_partition=$(kpartx -sav ${device} | grep "^add" | cut -d" " -f3)
+add_on_exit "kpartx -dv ${device}"
+
+loopback_dev="/dev/mapper/${device_partition}"
+
+# Mount partition
+image_mount_point=${work}/mnt
+mkdir -p ${image_mount_point}
+
+mount ${loopback_dev} ${image_mount_point}
+add_on_exit "umount ${image_mount_point}"
+
+# == Guide to variables in this script (all paths are defined relative to the real root dir, not the chroot)
+
+# work: the base working directory outside the chroot
+#      eg: /mnt/stemcells/aws/xen/centos/work/work
+# disk_image: path to the stemcell disk image
+#      eg: /mnt/stemcells/aws/xen/centos/work/work/aws-xen-centos.raw
+# device: path to the loopback devide mapped to the entire disk image
+#      eg: /dev/loop0
+# loopback_dev: device node mapped to the main partition in disk_image
+#      eg: /dev/mapper/loop0p1
+# image_mount_point: place where loopback_dev is mounted as a filesystem
+#      eg: /mnt/stemcells/aws/xen/centos/work/work/mnt
+
+# Generate random password
+random_password=$(tr -dc A-Za-z0-9_ < /dev/urandom | head -c 16)
+
+touch ${image_mount_point}${device}
+mount --bind ${device} ${image_mount_point}${device}
+add_on_exit "umount ${image_mount_point}${device}"
+
+mkdir -p `dirname ${image_mount_point}${loopback_dev}`
+touch ${image_mount_point}${loopback_dev}
+mount --bind ${loopback_dev} ${image_mount_point}${loopback_dev}
+add_on_exit "umount ${image_mount_point}${loopback_dev}"
+
+# GRUB 2 needs /sys and /proc to do its job
+mount -t proc none ${image_mount_point}/proc
+add_on_exit "umount ${image_mount_point}/proc"
+
+mount -t sysfs none ${image_mount_point}/sys
+add_on_exit "umount ${image_mount_point}/sys"
+
+echo "(hd0) ${device}" > ${image_mount_point}/device.map
+
+# install bootsector into disk image file
+run_in_chroot ${image_mount_point} "grub-install -v --no-floppy --grub-mkdevicemap=/device.map --target=i386-pc ${device}"
+
+# Enable password-less booting in openSUSE, only editing the boot menu needs to be restricted
+run_in_chroot ${image_mount_point} "sed -i 's/CLASS=\\\"--class gnu-linux --class gnu --class os\\\"/CLASS=\\\"--class gnu-linux --class gnu --class os --unrestricted\\\"/' /etc/grub.d/10_linux"
+cat >${image_mount_point}/etc/default/grub <<EOF
+GRUB_CMDLINE_LINUX="vconsole.keymap=us net.ifnames=0 biosdevname=0 crashkernel=auto selinux=0 plymouth.enable=0 console=ttyS0,115200n8 earlyprintk=ttyS0 rootdelay=300 ipv6.disable=1 audit=1 cgroup_enable=memory swapaccount=1 nvme_core.io_timeout=4294967295"
+EOF
```

Comment:
It seems unideal to copy this whole file to add a param, especially if we want to make more grub config changes in the future. Have you all tried appending the timeout param as a seperate ops file that runs next? It looks like the rest of this file is just setting up config.

Reply:
Indeed this was my first intention just to change the `GRUB_COMMAND_LINE`, call `grub-mkconfig` and place the script file after `image_install_grub`. The problem to me is that there will be some more changes on the generated files afterwards and I need to overtake more lines of code which makes it at the end confusing. So I decided to replaced the whole file for AWS. To introduce a new variable for the cloud provider seems also not to fit to the code structure.